pax-web-runtime/src/main/java/org/ops4j/pax/web/service/internal/ServletEventDispatcher.java

```diff
@@ -135,12 +134,6 @@ public void servletEvent(final ServletEvent event) {
 		}
 		synchronized (listeners) {
 			callListeners(event);
-			Map<String, ServletEvent> events = states.get(event.getBundle());
-			if (events == null) {
-				events = new LinkedHashMap<String, ServletEvent>();
-				states.put(event.getBundle(), events);
-			}
-			events.put(event.getAlias(), event);
```

Comment:
The event contains reference to bundles and servlets (in Event class) and are never removed. The time when new listener will be added is undefined. A new listener will not get all events, just last event having same bundle and alias (i. e. if listener will be added after UNDEPLOYED it may not get UNDEPLOYING event).

This indeed can be optimised by listening for bundle deactivation events, but in case of removing just one particular server reference to this server will be persisted.